### ISSUE TITLE
Unlabeled examples on -oaa changes the weights of the model

### ISSUE THREAD

**Issue by dahernan** [JohnLangford/vowpal_wabbit#1435]
I notice that making only predictions on -oaa changes the probabilities returned, is this a expected behavior?

This is the example I'm running (v8.5.0):
`vw --daemon --foreground --port 58009 --num_children 8 --oaa 2 --loss_function=logistic --probabilities --learning_rate=10 -b 24 --l1=1e-6 --l2=1e-6`

```
2 | size_small
> `1:0.500000 2:0.500000`
1 | size_large
> `1:0.163913 2:0.836087`
2 | size_small
> `1:0.203320 2:0.796680`
1 | size_large
> `1:0.864883 2:0.135117`
| size_large
>`1:0.953742 2:0.046258`
| size_small
> `1:0.008258 2:0.991742`
| size_small
> `1:0.054438 2:0.945562`
| size_small
> `1:0.104347 2:0.895653`
| size_small
> `1:0.142469 2:0.857532`
| size_small
> `1:0.171975 2:0.828025`
```
Notice how the probabilities are slowly balancing, even if you only do predictions, shouldn't stay constant?

**Comment by arielf** [comment]
This is good input.
 
Learn mixed with predict is very useful in many situations.  Yet `vw` seem to require `-t` to disable updates (i.e. a separate run).

John: wouldn't it make more sense (and be cool) cool if `vw` would only predict, and disable updates on examples which have no label?

IOW: disable the update step (imply `-t`) whenever a label is undefined.

This would make "learn vs with predict" friendlier to the user in many use-cases.

**Comment by JohnLangford** [comment]
I'm confused by the above: It looks like labeled examples are being fed in?  Why do you expect predictions when you don't specify -t and feed in labeled examples?

@arielf you can generally mix labeled and unlabeled examples.  Do you have an example to the contrary?

**Comment by dahernan** [comment]
@JohnLangford  using vw as a network service basically you can train a classifier online with labeled examples, meanwhile making predictions with unlabeled examples in what is been trained so far, that result very useful to me.

The problem comes when feeding unlabeled examples changes the value of the probabilities returned, it seems to me that a unlabeled example should not change the internal state

**Comment by JohnLangford** [comment]
Ah, I see it now.   This looks like a bug.  Can you reproduce it without daemon mode?

**Comment by arielf** [comment]
@JohnLangford, what @dahernan said:  I know you can mix unlabeled and labeled. 

However, the expectation is that unlabeled examples don't update/change the model weights.

Here's an example without `--daemon` mode. This is essentially, the same example David gave,
I just added `-a` so you can see how weights keep changing on the last two unlabeled examples.

```
$ vw -a --oaa 2 --loss_function=logistic --probabilities --learning_rate=10 -b 24 --l1=1e-6 --l2=1e-6
using l1 regularization = 1e-06
using l2 regularization = 1e-06
Num weight bits = 24
learning rate = 10
initial_t = 0
power_t = 0.5
using no cache
Reading datafile =
num sources = 1
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
2 | size_small
0
        size_small:12565722:1:0@0       Constant:6523576:1:0@0
0
        size_small:12565723:1:0@0       Constant:6523577:1:0@0
1.000000 1.000000            1            1.0        2   1(50%)        2
1 | size_large
-1.629397
        Constant:6523576:1:-1.6294@0.25 size_large:12439586:1:0@0
1.629400
        Constant:6523577:1:1.6294@0.25  size_large:12439587:1:0@0
1.000000 1.000000            2            2.0        1   2(84%)        2
2 | size_small
-1.365672
        size_small:12565722:1:-1.62938@0.25     Constant:6523576:1:0.263709@0.949041
1.365672
        size_small:12565723:1:1.62938@0.25      Constant:6523577:1:-0.263712@0.949042
1 | size_large
1.856449
        size_large:12439586:1:2.20551@0.699041  Constant:6523576:1:-0.349066@0.99038
-1.856458
        size_large:12439587:1:-2.20552@0.699042 Constant:6523577:1:0.349065@0.990381
0.500000 0.000000            4            4.0        1   1(86%)        2

# ---- now start sending unlabeled:
| size_large
3.026159
        size_large:12439586:1:2.83998@0.717298  Constant:6523576:1:0.186179@1.00864
-3.026169
        size_large:12439587:1:-2.83999@0.717299 Constant:6523577:1:-0.18618@1.00864
| size_large
-1.774355
        Constant:6523576:1:-2.11499@1.91826     size_large:12439586:1:0.340633@1.62692
-3.567896
        size_large:12439587:1:-3.13384@0.719438 Constant:6523577:1:-0.434055@1.01078
```

**Comment by arielf** [comment]
I can also reproduce without all the additional options.

`--oaa 2` is sufficient to reproduce changing weights on unlabeled examples.
Even entering empty lines/examples updates the weights.

Without `--oaa`, behavior is as expected.

**Comment by JohnLangford** [comment]
FYI, I'm working on this problem, but it's leading to cascading changes...

**Comment by dahernan** [comment]
No probs 👍, Thank you!

**Comment by matryer** [comment]
Same problem here. Would love a fix?

**Comment by JohnLangford** [comment]
I pushed a branch here:
https://github.com/JohnLangford/vowpal_wabbit/tree/fix_oaa_stateful_test_example
but some tests are failing so we need to root cause the test failures.

-John

**Comment by matryer** [comment]
Thanks John. I’ll see if that fixed it for us.

**Comment by JohnLangford** [comment]
I'd suggest caution here---we really need to root cause the test failures.

-John

**Comment by matryer** [comment]
Absolutely.

**Comment by dahernan** [comment]
I tried `fix_oaa_stateful_test_example` branch, and it fixes the problem, but it has a strange side effect when you save the model and load the model back again (all in daemon mode), it returns different predictions for the same dataset.

```
Training [15998/15998]Finished Training
Validation...

Predict [3999/3999] ok

Correct 3289 Incorrect 710, Acc 0.822456

Saving model...
file with the state /var/folders/tf/gk4g7f213pqg53512_tbb9wm0000gp/T/stateTest852901386
Restoring model...

Validation 2...
Predict [3999/3999] ok
Correct 3211 Incorrect 788, Acc 0.802951
```

**Comment by JohnLangford** [comment]
This patch has merged in now.  @dahernan if the problem you discussed still exists, can you make this reproducible?

**Comment by JohnLangford** [comment]
Incidentally: I wanted to mention that this ended up being a deep systemic change that revealed issues in multiple places, causing multiple tests to change.  The key change is here: 
https://github.com/JohnLangford/vowpal_wabbit/blob/master/vowpalwabbit/parser.cc#L817
I've made every label type have a test_label() function.    This is then dispatched by the learning loop here:
https://github.com/JohnLangford/vowpal_wabbit/blob/master/vowpalwabbit/global_data.cc#L146 

Which overall makes the semantics of -t, holdout examples, and test labels uniform across the codebase.

**Comment by dahernan** [comment]
Just tried master and it works good now, I'll try to run more test, and I'll raise a new issue if I found something, thank you very much!
